File: Python/2021/S3.py
# PASSED

people = [list(map(int, input().split())) for _ in range(int(input()))]

def calc(pos):
    output = 0

    for p, w, d in people:
        if abs(p - pos) > d:
            if p < pos:
                output += ((pos - d) - p) * w
            else:
                output += (p - (pos + d)) * w

    return output

start = 0
end = max(p[0] for p in people)

while start + 1 < end:
    mid = (start + end) // 2

    time1, time2 = calc(mid), calc(mid + 1)

    if time1 >= time2:
        start = mid
    else:
        end = mid

print(min(calc(start), calc(end)))

File: C++/2021/S3.cpp
#include <bits/stdc++.h>

using namespace std;

int main() {  
  cin.tie(0);
  int low = 0, high = 0, n;cin >> n;
  vector<int> p(n), w(n), d(n);
  for (int i = 0; i < n; i++) {
    cin >> p[i] >> w[i] >> d[i];
    high = max(high, p[i]);
  }
  auto check = [&](int pos) {
    long long sum = 0;
    for (int i = 0; i < n; i++) {
        if (p[i] + d[i] < pos) sum += (long long) abs(p[i] - (pos - d[i])) * w[i];
        else if (p[i] - d[i] > pos) sum += (long long) abs(p[i] - (pos + d[i])) * w[i];
    }
    return sum;
  };  

  while (low + 1 < high) {
    int mid = (low + high) >> 1;  
    if (check(mid) < check(mid - 1)) low = mid + 1;
    else high = mid - 1;
  } 
  cout << min(check(low), check(high)) << '\n';
  return 0;
}